config/config.go

```diff
@@ -1,675 +1,208 @@
 )
 
-var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+// regular expression matching lowecase letter followed by an uppercase letter
+var camelCaseMatch = regexp.MustCompile("([a-z0-9])([A-Z])")
 
-var (
-	whSchemaVersion        string
-	hotReloadableConfig    map[string][]*ConfigVar
-	nonHotReloadableConfig map[string][]*ConfigVar
-	configVarLock          sync.RWMutex
-)
-
-type ConfigVar struct {
-	value           interface{}
-	multiplier      interface{}
-	isHotReloadable bool
-	defaultValue    interface{}
-	keys            []string
-}
-
-func newConfigVar(value, multiplier, defaultValue interface{}, isHotReloadable bool, keys []string) *ConfigVar {
-	return &ConfigVar{
-		value:           value,
-		multiplier:      multiplier,
-		isHotReloadable: isHotReloadable,
-		defaultValue:    defaultValue,
-		keys:            keys,
-	}
-}
+// regular expression matching uppercase letters contained in environment variable names
+var upperCaseMatch = regexp.MustCompile("^[A-Z0-9_]+$")
 
-func getConfigValue(configVar *ConfigVar) interface{} {
-	return configVar.value
-}
-
-// Rudder server supported config constants
-const (
-	EmbeddedMode      = "embedded"
-	MasterMode        = "master"
-	MasterSlaveMode   = "master_and_slave"
-	SlaveMode         = "slave"
-	OffMode           = "off"
-	PooledWHSlaveMode = "embedded_master"
-)
-
-// TransformKey as package method to get the formatted env from a give string
-func TransformKey(s string) string {
-	snake := matchAllCap.ReplaceAllString(s, "${1}_${2}")
-	snake = strings.ReplaceAll(snake, ".", "_")
-	return "RSERVER_" + strings.ToUpper(snake)
-}
+// default, singleton config instance
+var c *Config
 
-func Load() {
-	if err := godotenv.Load(); err != nil {
-		fmt.Println("INFO: No .env file found.")
-	}
-	hotReloadableConfig = make(map[string][]*ConfigVar)
-	nonHotReloadableConfig = make(map[string][]*ConfigVar)
-	configPath := GetEnv("CONFIG_PATH", "./config/config.yaml")
-	viper.SetConfigFile(configPath)
-	err := viper.ReadInConfig() // Find and read the config file
-	UpdateConfig()
-	// Don't panic if config.yaml is not found or error with parsing. Use the default config values instead
-	if err != nil {
-		fmt.Println("[Config] :: Failed to parse Config Yaml, using default values:", err)
-	}
+func init() {
+	c = New()
 }
 
-func UpdateConfig() {
-	viper.WatchConfig()
-	viper.OnConfigChange(func(e fsnotify.Event) {
-		watchForConfigChange()
-	})
+// Reset resets configuration
+func Reset() {
+	c = New()
 }
 
-func watchForConfigChange() {
-	defer func() {
-		if r := recover(); r != nil {
-			err := fmt.Errorf("cannot update Config Variables: %v", r)
-			fmt.Println(err)
-		}
-	}()
-	configVarLock.RLock()
-	defer configVarLock.RUnlock()
-	_ = checkAndHotReloadConfig(hotReloadableConfig)
-	isChanged := checkAndHotReloadConfig(nonHotReloadableConfig)
-	if isChanged && GetEnvAsBool("RESTART_ON_CONFIG_CHANGE", false) {
-		os.Exit(1)
-	}
+// New creates a new config instance
+func New() *Config {
```

Comment:
If we want to stick to the singleton pattern, do we even need this constructor? It also seems to be unused, perhaps we can remove it and just stick to the fact that we'll call the functions (e.g. `func GetBool`) instead of the methods (e.g. `func (c *Config) GetBool`).

Reply:
the singleton pattern and package functions are there only because it is not possible at this stage to proceed without them. The goal is to move away from a singleton config instance, for the reasons we have already discussed (hard to test, parallelise, etc. )

Reply:
We are not moving away from it now, in fact the constructor is not used. I'd refrain from adding code meant for future use until needed.

Anyway, can you explain what is blocking us from using `config` only during bootstrapping and removing it from each package?

In my head the ideal scenario is to stop spreading `config` in all packages and read the configuration in places like `main.go`, app handlers etcetera only. Then pass the read data as parameters when instantiating things. That way we remove the need of having something thread-safe and we decouple from config altogether. What do you think are the blockers to go towards that? Hot reloads?

Reply:
1. the `New` function is being used already by `config.init`, `config.Reset` & `config_tests`.
2. We need to refactor a lot of codebase to stop using the config package as we are doing now. Codebase in packages that are already using global variables for their configuration purposes. IMO this is something that needs to be done gradually and with care, not in bulk, since each case might have its own unique details we need to pay attention to. This also involves getting rid of the [various init functions](https://github.com/rudderlabs/rudder-server/blob/master/main.go#L140) that we have hanging around :)

Reply:
Agreed :+1: In regards to `1` I know but I mentiond it because those are all within the `config` package. In any case, let's tackle this in the long run, it is a sensible approach :+1: